error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/TypeMemberOrdering.java

```diff
@@ -0,0 +1,236 @@
+                        .anyMatch(RECOGNIZED_SUPPRESSIONS::contains))
+        .orElse(false);
+  }
+
+  private static SuggestedFix replaceTypeMembers(
+      ImmutableList<TypeMember> typeMembers,
+      ImmutableList<TypeMember> replacementTypeMembers,
+      VisitorState state) {
+    return Streams.zip(
+            typeMembers.stream(),
+            replacementTypeMembers.stream(),
+            (original, replacement) -> replaceTypeMember(original, replacement, state))
+        .reduce(SuggestedFix.builder(), SuggestedFix.Builder::merge, SuggestedFix.Builder::merge)
+        .build();
+  }
+
+  private static SuggestedFix replaceTypeMember(
+      TypeMember original, TypeMember replacement, VisitorState state) {
+    /* Technically this check is not necessary, but it avoids redundant replacements. */
+    if (original.equals(replacement)) {
+      return SuggestedFix.emptyFix();
+    }
+
+    String replacementSource =
+        Stream.concat(
+                replacement.comments().stream(),
+                Stream.of(SourceCode.treeToString(replacement.tree(), state)))
+            .collect(joining(System.lineSeparator()));
+    return SuggestedFixes.replaceIncludingComments(
+        TreePath.getPath(state.getPath(), original.tree()), replacementSource, state);
+  }
+
+  /** Returns the type's members with their comments. */
+  private static ImmutableList<TypeMember> getTypeMembersWithComments(
+      ClassTree tree, VisitorState state) {
+    return tree.getMembers().stream()
+        .map(
+            member ->
+                new AutoValue_TypeMemberOrdering_TypeMember(
+                    member, getTypeMemberComments(tree, member, state)))
+        .collect(toImmutableList());
+  }
+
+  private static ImmutableList<String> getTypeMemberComments(
+      ClassTree tree, Tree member, VisitorState state) {
+    int typeStart = ASTHelpers.getStartPosition(tree);
+    int typeEnd = state.getEndPosition(tree);
+    int memberStart = ASTHelpers.getStartPosition(member);
+    int memberEnd = state.getEndPosition(member);
+    if (typeStart == Position.NOPOS
+        || typeEnd == Position.NOPOS
+        || memberStart == Position.NOPOS
+        || memberEnd == Position.NOPOS) {
+      /* Source code details appear to be unavailable. */
+      return ImmutableList.of();
+    }
+
+    // TODO: Move identifying "previous member end position" to an outer loop,
+    //  Loop once and identify for all members
+    // TODO: Check if this handles properly comments on the first member.
+    Optional<Integer> previousMemberEndPos =
+        state.getOffsetTokens(typeStart, typeEnd).stream()
+            .map(ErrorProneToken::endPos)
+            .takeWhile(endPos -> endPos < memberStart)
+            .reduce((earlierPos, laterPos) -> laterPos);
+
+    List<ErrorProneToken> typeMemberTokens =
+        state.getOffsetTokens(previousMemberEndPos.orElse(memberStart), memberEnd);
```

Comment:
This TODO is for addressing the very not performant ways of this solution.

The challenge here to keep the comment and maybe other code too between members.

Options (I can think of atm):

---

- Somehow get all `ErrorProneTokens` (or source) of the class,
- divide them for each member, and
- reorder these token+member groups.

I believe this was described in this [comment](https://github.com/PicnicSupermarket/error-prone-support/pull/636#discussion_r1362763320).

With this I'm not sure how we can handle the whitespace properly, but I'm yet to fully discover this option.

---

- Go through the type's tokens once,
- gather all commented tokens, and put them in a `Map<Identifier, Comments>`
- use this map to get the comments of a member (`tree.getMembers()`)

I tried this quickly, using `ErrorProneToken#pos` and `ASTHelpers.getStartPosition(member)` as identifier.

This worked in general, but failed the `detached comment from method` test and the `replacementSuggestedFixDoesNotModifyWhitespace`.

---
---

Additional Note:

`VisitorState#getOffsetTokens` java doc says:
> This is moderately expensive (the source of the node has to be re-lexed), so it should only be used if a fix is already going to be emitted.

With this in mind, we should first determine if the members are ordered properly, and only then fetch the comments for them - if we opt to use this helper.